README.md

[![Arduino CI](https://github.com/RobTillaart/Max44007/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/Max44007/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/Max44007/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/Max44007/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/Max44007/actions/workflows/jsoncheck.yml)
[![GitHub issues](https://img.shields.io/github/issues/RobTillaart/Max44007.svg)](https://github.com/RobTillaart/Max44007/issues)

[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/Max44007/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/Max44007.svg?maxAge=3600)](https://github.com/RobTillaart/Max44007/releases)
[![PlatformIO Registry](https://badges.registry.platformio.org/packages/robtillaart/library/Max44007.svg)](https://registry.platformio.org/libraries/robtillaart/Max44007)


# Max44007 I2C LUX sensor

Library for Max44007 I2C lux sensor.


## Description

The Max44007 ambient light sensor is an I2C sensor, that has a 22 bit 
dynamic range from 0.025 lux to 104,448 lux.

This library is experimental and created by changing some key values in the
very similar MAX44009 library. As there was no hardware available the library 
is not tested yet. If you do, please share the results (on GitHub)

Note: version 0.2.0 simplified the constructor interface and removed **configure()**.


## Schema breakout Max44007


```cpp
//      breakout Max44007
//
//          +--------+
//      VCC |o       |
//      GND |o       |
//      SCL |o      o| ADDRESS
//      SDA |o      o| -INT
//          +--------+
//
//  ADDRESS:
//  0 = 0x5A
//  1 = 0x5B
//
//  INT:
//  Connect the INT pin to an pull up resistor
//  0 = interrupt
//  1 = no interrupt
//  connect to an LED or an interrupt pin of an Arduino
//
```


## Interface

```cpp
#include "Max44007.h"
```

### Constructor 
  
- **Max44007(const uint8_t address = MAX44007_DEFAULT_ADDRESS, TwoWire \*wire = &Wire)** Constructor.
Optional address and optional I2C interface.\
- **bool isConnected()** returns true if the device address configured is available on I2C bus.

NOTE: The user must call **Wire.begin()** or **Wire.begin(SDA, SCL)** in **setup()**.


### Basic 

- **float getLux()** read the sensor and return the value in LUX. If the value is negative, an error has occurred. 
- **int getError()** returns last error.

```cpp
//  ERROR CODES
#define Max44007_OK                     0
#define Max44007_ERROR_WIRE_REQUEST    -10
#define Max44007_ERROR_OVERFLOW        -20
#define Max44007_ERROR_HIGH_BYTE       -30
#define Max44007_ERROR_LOW_BYTE        -31
```


### Configure thresholds

check datasheet for details

- **bool setHighThreshold(float value)** sets the upper threshold for the interrupt 
generation (INT pulls LOW). Works only if INTE bit is set by **enableInterrupt()**. 
Function returns false if the value is out of range.
- **float getHighThreshold()** returns the value set.
- **bool setLowThreshold(float value)** sets the lower threshold for the interrupt 
generation (INT pulls LOW). Works only if INTE bit is set by **enableInterrupt()**. 
Function returns false if the value is out of range.
- **float getLowThreshold()** returns the value set.
- **void setThresholdTimer(uint8_t value)** Time the threshold needs to be exceeded, 
defined in steps of 100ms. 2 seems to be a practical minimum.
- **uint8_t getThresholdTimer()** returns the value set.


### Configure interrupts

check datasheet for details

- **void enableInterrupt()** enables the upper and lower threshold interrupts.
- **void disableInterrupt()** disables the threshold interrupts.
- **bool interruptEnabled()** returns 1 if the interrupt mode is enabled.
- **uint8_t getInterruptStatus()** returns 1 if an interrupt has occurred.


### Configure flags

check datasheet for details

- **void setConfiguration(uint8_t)** writes directly to configuration register. 
**warning** Use with care.
- **uint8_t getConfiguration()** reads the current configuration register.


### Configure sample mode

check datasheet for details

CCR = Current Divisor Ratio.

TIM = Integration time.

- **void setAutomaticMode()** in automatic mode the Max44007 determines the CDR and TIM 
parameters.
- **void setContinuousMode()** continuous mode uses more power than a "single" conversion. 
Advantage is that the latest data is always available fast.
- **void clrContinuousMode()** uses less power so better for LOW power configurations. 
- **void setManualMode(uint8_t CDR, uint8_t TIM)** Set the Current Divisor Ratio and the 
integration time manually. Effectively disable automatic mode.
- **int getIntegrationTime()** returns the set integration time in milliseconds

```
    CDR = Current Divisor Ratio
    CDR = 1 ==> only 1/8th is measured

    TIM = Time Integration Measurement (table)
    000   800ms
    001   400ms
    010   200ms
    011   100ms
    100    50ms       manual only
    101    25ms       manual only
    110    12.5ms     manual only
    111     6.25ms    manual only
```


### Test functions

Function for the conversion math, not meant to be used directly, 
but by making them public they become testable.

- **float convertToLux(uint8_t dataHigh, uint8_t dataLow)** convert intern register 
format to a LUX value. 


## Examples

**Max44007\examples\Max44007_test01**
- use for e.g. UNO (use voltage convertor!)

**Max44007\examples\Max44007_test02**
- will **not** compile for UNO
- use for e.g. ESP32


## Notes

Please be aware this is a **3.3 Volt device** so it should not be connected
to an Arduino or other 5 Volt device directly. Use a level convertor to 
solve this.

Do not forget to connect the address pin as you cannot read the sensor
in a reliable way. As the line will float it will sometimes have the
right address and sometimes not. (been there ;)

Pull ups on I2C bus are recommended.


## Future

#### Must

- follow MAX44009

#### Should

#### Could

- Merge MAX44007 into MAX44009 class hierarchy.

#### Wont


## Support

If you appreciate my libraries, you can support the development and maintenance.
Improve the quality of the libraries by providing issues and Pull Requests, or
donate through PayPal or GitHub sponsors.

Thank you,



CHANGELOG.md
# Change Log MAX44007

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.2.0] - 2023-10-02
- refactor constructor around I2C interface (simpler)
- remove **configure()**
- update readme.md

----

## [0.1.2] - 2022-11-16
- add RP2040 in build-CI
- add changelog.md

## [0.1.1] - 2022-01-05
- minor refactor

## [0.1.0] - 2022-01-04
- initial version (derived from Max44009 0.5.1)




test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-30
// PURPOSE: unit tests for the Max44007
//          https://github.com/RobTillaart/Max44007
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "Max44007.h"


unittest_setup()
{
  fprintf(stderr, "Max44007_LIB_VERSION: %s\n", (char *) Max44007_LIB_VERSION);
}


unittest_teardown()
{
}


unittest(test_constants)
{
  assertEqual(Max44007_DEFAULT_ADDRESS, 0x5A);
  assertEqual(Max44007_ALT_ADDRESS    , 0x5B);

  assertEqual(Max44007_INTERRUPT_STATUS  , 0x00);
  assertEqual(Max44007_INTERRUPT_ENABLE  , 0x01);
  assertEqual(Max44007_CONFIGURATION     , 0x02);
  assertEqual(Max44007_LUX_READING_HIGH  , 0x03);
  assertEqual(Max44007_LUX_READING_LOW   , 0x04);
  assertEqual(Max44007_THRESHOLD_HIGH    , 0x05);
  assertEqual(Max44007_THRESHOLD_LOW     , 0x06);
  assertEqual(Max44007_THRESHOLD_TIMER   , 0x07);

  assertEqual(Max44007_CFG_CONTINUOUS    , 0x80);
  assertEqual(Max44007_CFG_MANUAL        , 0x40);
  assertEqual(Max44007_CFG_CDR           , 0x08);
  assertEqual(Max44007_CFG_TIMER         , 0x07);

  assertEqual(Max44007_OK                , 0  );
  assertEqual(Max44007_ERROR_WIRE_REQUEST, -10);
  assertEqual(Max44007_ERROR_OVERFLOW    , -20);
  assertEqual(Max44007_ERROR_HIGH_BYTE   , -30);
  assertEqual(Max44007_ERROR_LOW_BYTE    , -31);
}


unittest(test_constructor)
{
  Max44007 LuxA(0x5A);
  Wire.begin();

  assertEqual(Max44007_OK, LuxA.getError());
  assertTrue(LuxA.isConnected());   // TODO should be false...

  // TODO more tests if WIRE works...
}


unittest(test_convertToLux)
{
  Max44007 LuxA(0x5A);

  assertEqualFloat(0.000, LuxA.convertToLux(0x00, 0x00), 0.0001);
  assertEqualFloat(0.025, LuxA.convertToLux(0x00, 0x01), 0.0001);
  assertEqualFloat(0.400, LuxA.convertToLux(0x01, 0x00), 0.0001);
  assertEqualFloat(0.850, LuxA.convertToLux(0x11, 0x01), 0.0001);
  fprintf(stderr, "\n");

  assertEqualFloat(103630, LuxA.convertToLux(0xEF, 0x0E), 1);
  assertEqualFloat(104448, LuxA.convertToLux(0xEF, 0x0F), 1);
  fprintf(stderr, "\ndone...\n");
}


unittest_main()

// --------


Max44007.cpp
//
//    FILE: Max44007.cpp
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.0
// PURPOSE: library for Max44007 lux sensor Arduino
//     URL: https://github.com/RobTillaart/Max44007
//
//  HISTORY
//
//  0.1.0   2022-01-04  initial version (derived from Max44009 0.5.1)


#include "Max44007.h"


// MAX44007 KEY VALUES
#define MAX44007_MIN_LUX                       (0.025)
#define MAX44007_MAX_LUX                       (104448.0)



#if defined(ESP8266) || defined(ESP32)
Max44007::Max44007(const uint8_t address, const uint8_t dataPin, const uint8_t clockPin)
{
  _address = address;
  _data    = 0;
  _error   = Max44007_OK;
  _wire    = &Wire;

  if ((dataPin < 255) && (clockPin < 255))
  {
    _wire->begin(dataPin, clockPin);
  } else {
    _wire->begin();
  }
}
#endif


Max44007::Max44007(const uint8_t address, const Boolean begin)
{
  Max44007::configure(address, &Wire, begin);
}


Max44007::Max44007(const Boolean begin)
{
  Max44007::configure(Max44007_DEFAULT_ADDRESS, &Wire, begin);
}


void Max44007::configure(const uint8_t address, TwoWire *wire, const Boolean begin)
{
  _address = address;
  _data    = 0;
  _error   = Max44007_OK;
  _wire    = wire;

  if (begin == Boolean::True)
  {
    _wire->begin();
  }
}


bool Max44007::isConnected()
{
  _wire->beginTransmission(_address);
  _error = _wire->endTransmission();
  return (_error == 0);
}


float Max44007::getLux(void)
{
  uint8_t datahigh = read(Max44007_LUX_READING_HIGH);
  if (_error != Max44007_OK)
  {
    _error = Max44007_ERROR_HIGH_BYTE;
    return _error;
  }
  uint8_t datalow = read(Max44007_LUX_READING_LOW);
  if (_error != Max44007_OK)
  {
    _error = Max44007_ERROR_LOW_BYTE;
    return _error;
  }
  uint8_t exponent = datahigh >> 4;
  if (exponent == 0x0F)
  {
    _error = Max44007_ERROR_OVERFLOW;
    return _error;
  }

  float lux = convertToLux(datahigh, datalow);
  return lux;
}


int Max44007::getError()
{
  int err = _error;
  _error = Max44007_OK;
  return err;
}


bool Max44007::setHighThreshold(const float value)
{
  return setThreshold(Max44007_THRESHOLD_HIGH, value);
}


float Max44007::getHighThreshold(void)
{
  return getThreshold(Max44007_THRESHOLD_HIGH);
}


bool Max44007::setLowThreshold(const float value)
{
  return setThreshold(Max44007_THRESHOLD_LOW, value);
}


float Max44007::getLowThreshold(void)
{
  return getThreshold(Max44007_THRESHOLD_LOW);
}


void Max44007::setThresholdTimer(const uint8_t value)
{
  write(Max44007_THRESHOLD_TIMER, value);
}


uint8_t Max44007::getThresholdTimer()
{
  return read(Max44007_THRESHOLD_TIMER);
}


void Max44007::setConfiguration(const uint8_t value)
{
  write(Max44007_CONFIGURATION, value);
}


uint8_t Max44007::getConfiguration()
{
  return read(Max44007_CONFIGURATION);
}


void Max44007::setAutomaticMode()
{
  // CDR & TIM cannot be written in automatic mode
  uint8_t config = read(Max44007_CONFIGURATION);
  config &= ~Max44007_CFG_MANUAL;
  write(Max44007_CONFIGURATION, config);
}


void Max44007::setContinuousMode()
{
  uint8_t config = read(Max44007_CONFIGURATION);
  config |= Max44007_CFG_CONTINUOUS;
  write(Max44007_CONFIGURATION, config);
}


void Max44007::clrContinuousMode()
{
  uint8_t config = read(Max44007_CONFIGURATION);
  config &= ~Max44007_CFG_CONTINUOUS;
  write(Max44007_CONFIGURATION, config);
}


void Max44007::setManualMode(uint8_t CDR, uint8_t TIM)
{
  if (CDR !=0) CDR = 1;    // only 0 or 1 
  if (TIM > 7) TIM = 7;
  uint8_t config = read(Max44007_CONFIGURATION);
  config |= Max44007_CFG_MANUAL;
  config &= 0xF0;                     // clear old CDR & TIM bits
  config |= CDR << 3 | TIM;           // set new CDR & TIM bits
  write(Max44007_CONFIGURATION, config);
}


float Max44007::convertToLux(uint8_t datahigh, uint8_t datalow)
{
  uint8_t  exponent = datahigh >> 4;
  uint32_t mantissa = ((datahigh & 0x0F) << 4) + (datalow & 0x0F);
  float lux = ((0x0001 << exponent) * MAX4407_MIN_LUX) * mantissa;
  return lux;
}


///////////////////////////////////////////////////////////
//
// PRIVATE
//
bool Max44007::setThreshold(const uint8_t reg, const float value)
{
  // CHECK RANGE OF VALUE
  if ((value < 0.0) || (value > MAX44007_MAX_LUX)) return false;

  uint32_t mantissa = round(value * (1.0 / MAX44007_MIN_LUX));     //  compile time optimized.
  uint8_t exponent = 0;
  while (mantissa > 255)
  {
    mantissa >>= 1;                // bits get lost
    exponent++;
  };
  mantissa = (mantissa >> 4) & 0x0F;
  exponent <<= 4;
  write(reg, exponent | mantissa);
  return true;
}


float Max44007::getThreshold(uint8_t reg)
{
  uint8_t datahigh = read(reg);
  float lux = convertToLux(datahigh, 0x08);  // 0x08 = correction for lost bits 
  return lux;
}


uint8_t Max44007::read(uint8_t reg)
{
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _error = _wire->endTransmission();
  if (_error != Max44007_OK)
  {
    return _data; // last value
  }
  if (_wire->requestFrom(_address, (uint8_t) 1) != 1)
  {
    _error = Max44007_ERROR_WIRE_REQUEST;
    return _data; // last value
  }
  _data = _wire->read();
  return _data;
}


void Max44007::write(uint8_t reg, uint8_t value)
{
  _wire->beginTransmission(_address);
  _wire->write(reg);
  _wire->write(value);
  _error = _wire->endTransmission();
}


// -- END OF FILE --


Max44007.h
#pragma once

//    FILE: Max44007.h
//  AUTHOR: Rob Tillaart
// VERSION: 0.1.2
// PURPOSE: library for Max44007 lux sensor Arduino
// HISTORY: See Max44007.cpp


//       breakout MAX44007
//
//          +--------+
//      VCC |o       |
//      GND |o       |
//      SCL |o      o| ADDRESS
//      SDA |o      o| -INT
//          +--------+
//
//  ADDRESS:
//  0 = 0x4A
//  1 = 0x4B
//
//  INT:
//  Connect the INT pin to an pull up resistor
//  0 = interrupt
//  1 = no interrupt
//  connect to an LED or an interrupt pin of an Arduino
//


#include "Wire.h"
#include "Arduino.h"


#define MAX44007_LIB_VERSION                  (F("0.1.2"))

#define MAX44007_DEFAULT_ADDRESS              0x5A
#define MAX44007_ALT_ADDRESS                  0x5B


//  REGISTERS
#define MAX44007_INTERRUPT_STATUS             0x00
#define MAX44007_INTERRUPT_ENABLE             0x01
#define MAX44007_CONFIGURATION                0x02
#define MAX44007_LUX_READING_HIGH             0x03
#define MAX44007_LUX_READING_LOW              0x04
#define MAX44007_THRESHOLD_HIGH               0x05
#define MAX44007_THRESHOLD_LOW                0x06
#define MAX44007_THRESHOLD_TIMER              0x07


//  CONFIGURATION MASKS
#define MAX44007_CFG_CONTINUOUS               0x80
#define MAX44007_CFG_MANUAL                   0x40
#define MAX44007_CFG_CDR                      0x08
#define MAX44007_CFG_TIMER                    0x07


//  ERROR CODES
#define MAX44007_OK                           0
#define MAX44007_ERROR_WIRE_REQUEST           -10
#define MAX44007_ERROR_OVERFLOW               -20
#define MAX44007_ERROR_HIGH_BYTE              -30
#define MAX44007_ERROR_LOW_BYTE               -31



class Max44007
{
public:
  //  enum class to prevent bool to be implicitly casted to int
  enum class Boolean { True, False };

#if defined(ESP8266) || defined(ESP32)
  //  dataPin and clockPin can be used for ESP8266
  Max44007(const uint8_t address, const uint8_t dataPin, const uint8_t clockPin);
#endif

  //  constructor for UNO
  Max44007(const uint8_t address, const Boolean begin = Boolean::True);
  //  default constructor with default address
  Max44007(const Boolean begin = Boolean::True);
  //  Change I2C interface and address
  void configure(const uint8_t address, TwoWire *wire, const Boolean begin = Boolean::True);

  bool    isConnected();
  float   getLux();
  int     getError();

  //  threshold must be between 0 and 188006
  bool    setHighThreshold(const float value);       //  returns false if value out of range
  float   getHighThreshold(void);
  bool    setLowThreshold(const float value);        //  returns false if value out of range
  float   getLowThreshold(void);
  void    setThresholdTimer(const uint8_t value);    / / 2 seems practical minimum
  uint8_t getThresholdTimer();

  void    enableInterrupt()    { write(MAX44007_INTERRUPT_ENABLE, 1); };
  void    disableInterrupt()   { write(MAX44007_INTERRUPT_ENABLE, 0); };
  bool    interruptEnabled()   { return read(MAX44007_INTERRUPT_ENABLE) & 0x01; };
  uint8_t getInterruptStatus() { return read(MAX44007_INTERRUPT_STATUS) & 0x01; };

  //  check datasheet for detailed behavior
  void    setConfiguration(uint8_t);
  uint8_t getConfiguration();
  void    setAutomaticMode();
  void    setContinuousMode();    //  uses more power
  void    clrContinuousMode();    //  uses less power
  //  CDR = Current Divisor Ratio
  //  CDR = 1 ==> only 1/8th is measured
  //  TIM = Time Integration Measurement (table)
  //  000   800ms
  //  001   400ms
  //  010   200ms
  //  011   100ms
  //  100    50ms       manual only
  //  101    25ms       manual only
  //  110    12.5ms     manual only
  //  111     6.25ms    manual only
  void    setManualMode(uint8_t CDR, uint8_t TIM);
  int     getIntegrationTime() { return 800 >> (getConfiguration() & 0x07); };  //  ms

  //  TEST the math
  float   convertToLux(uint8_t datahigh, uint8_t datalow);


protected:
  bool    setThreshold(uint8_t reg, float value);
  float   getThreshold(uint8_t reg);

  uint8_t read(uint8_t reg);
  void    write(uint8_t reg, uint8_t value);

  uint8_t _address;
  uint8_t _data;
  int     _error;

  TwoWire* _wire;
};


// -- END OF FILE --

